contracts/payment/LemonadeEscrowV1.sol

```diff
@@ -0,0 +1,226 @@
+        RefundPolicy[] memory refundPolicies
+    ) PaymentSplitter(payees, shares) DelegateManager(delegates) {
+        require(hostRefundPercent <= 100, "Invalid hostRefundPercent");
+
+        //-- check valid refundPolicies
+        if (refundPolicies.length == 1) {
+            _assertValidRefundPercent(refundPolicies[0]);
+        } else if (refundPolicies.length > 1) {
+            _assertValidRefundPercent(refundPolicies[0]);
+
+            for (uint16 i = 0; i < refundPolicies.length - 1; i++) {
+                require(
+                    refundPolicies[i].timestamp <
+                        refundPolicies[i + 1].timestamp &&
+                        refundPolicies[i].percent >
+                        refundPolicies[i + 1].percent,
+                    "Invalid refund policy order & percent"
+                );
+
+                _assertValidRefundPercent(refundPolicies[i + 1]);
+            }
+        }
+
+        _hostRefundPercent = hostRefundPercent;
+
+        for (uint16 i = 0; i < refundPolicies.length; i++) {
+            _refundPolicies.push(refundPolicies[i]);
+        }
+
+        _transferOwnership(owner);
+    }
+
+    //- modifiers
+
+    modifier onlyOwnerOrDelegate() {
+        require(
+            msg.sender == owner() || isDelegate(msg.sender),
+            "Must be owner or delegate"
+        );
+        _;
+    }
+
+    modifier escrowOpen() {
+        require(!_closed, "Escrow had been closed");
+        _;
+    }
+
+    //-- public write functions
+
+    function addDelegates(
+        address[] memory addresses
+    ) public override onlyOwner {
+        return super.addDelegates(addresses);
+    }
+
+    function removeDelegates(
+        address[] memory addresses
+    ) public override onlyOwner {
+        return super.removeDelegates(addresses);
+    }
+
+    function deposit(
+        uint256 paymentId,
+        address token,
+        uint128 amount
+    ) external payable override escrowOpen {
+        require(!_paymentCancelled[paymentId], "Payment had been cancelled");
+        require(amount > 0, "Amount must not be zero");
+
+        if (token == address(0)) {
+            require(msg.value == amount, "Amount not matched");
+        } else {
+            IERC20(token).transferFrom(msg.sender, address(this), amount);
+        }
+
+        _deposits[msg.sender][paymentId].push(Deposit(token, amount));
+
+        emit GuestDeposit(msg.sender, paymentId, token, amount);
+    }
+
+    function cancelByGuest(uint256 paymentId) external override escrowOpen {
+        require(
+            !_paymentCancelled[paymentId],
+            "Payment had already been cancelled"
+        );
+
+        //-- calculate refund percent based on policy
+        uint16 percent = 0;
+
+        if (_refundPolicies.length > 0) {
+            for (uint16 i = 0; i < _refundPolicies.length; i++) {
+                RefundPolicy memory policy = _refundPolicies[
+                    _refundPolicies.length - 1 - i
```

Comment:
I wanted a backward for loop, but Solidity throw error when iterator reaches -1, hence the forward loop with this index calculation